File: IVI_ECU/app/src/main/java/com/hackathon/v2x/ivi/model/R4Message.kt
package com.hackathon.v2x.ivi.model

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json

/**
 * Sealed Kotlin binding of the frozen R4 ADA-to-IVI message set (4.0.1.5,
 * `contracts/r4-ada-ivi.schema.json`): an [R4WarningEvent] or an
 * [R4StateMessage], discriminated on the wire by `"type": "warning" | "state"`
 * (see [R4Json]).
 */
@Serializable
sealed class R4Message {
    abstract val schemaVersion: Int
}

/**
 * Edge-triggered warning event: carries the R3 snapshot of the triggering
 * track (frozen R3 schema 3.0.1.4, `$ref` from the R4 schema) and the composed
 * [SceneGeometry].
 */
@Serializable
@SerialName("warning")
data class R4WarningEvent(
    override val schemaVersion: Int,
    val warningType: String,
    val riskState: String,
    @SerialName("object") val objectSnapshot: R3Snapshot,
    val geometry: SceneGeometry,
) : R4Message() {

    val trackedObject: R3Snapshot get() = objectSnapshot

    companion object {
        const val TYPE_KEY = "warning"
        const val WARNING_TYPE_NLOS_OBSTRUCTION = "nlos_obstruction"
        const val UNKNOWN_WARNING_TYPE = "unknown"
    }
}

/** Periodic awareness state (R15), last-value-wins by [seq]. */
@Serializable
@SerialName("state")
data class R4StateMessage(
    override val schemaVersion: Int,
    val seq: Long,
    val vehicles: R4Vehicles,
) : R4Message() {
    companion object {
        const val TYPE_KEY = "state"
    }
}

/** Minimal vehicle state carried inside the R4 state heartbeat. */
@Serializable
data class VehicleState(
    @SerialName("position") val position: VehiclePosition,
    @SerialName("speed") val speed: Float,
)

/** Vehicle positions of the state message; [vehicleC] absent-or-null until C is relayed. */
@Serializable
data class R4Vehicles(
    val ego: VehiclePosition,
    val vehicleB: VehiclePosition,
    val vehicleC: VehiclePosition? = null,
)

/** Binding-level Json for the R4 wire: "type" discriminates warning|state; unknown fields are ignored (additive evolution, HLD D4). */
val R4Json: Json = Json {
    classDiscriminator = "type"
    ignoreUnknownKeys = true
}

File: ADA_ECU/src/contracts/r4_message.cpp
#include "contracts/r4_message.hpp"

namespace ada::contracts {

void to_json(nlohmann::json& j, const R4VehicleSet& v) {
  j = nlohmann::json{{"ego", v.ego}, {"vehicleB", v.vehicleB}};
  // vehicleC key always emitted, null when nullopt — valid for both the
  // warning's required-nullable `geometry.vehicleC` and the state's
  // absent-or-null `vehicles.vehicleC` (see header comment).
  j["vehicleC"] = v.vehicleC.has_value() ? nlohmann::json(*v.vehicleC) : nlohmann::json(nullptr);
}

void from_json(const nlohmann::json& j, R4VehicleSet& v) {
  // j.at(...) throws on missing keys (schema-required); unknown extra keys
  // are deliberately ignored for additive evolution.
  j.at("ego").get_to(v.ego);
  j.at("vehicleB").get_to(v.vehicleB);
  // Missing key OR null both map to nullopt (state allows absence, warning
  // requires the key but allows null).
  const auto it = j.find("vehicleC");
  v.vehicleC = (it == j.end() || it->is_null()) ? std::nullopt
                                                : std::optional<Vec2>(it->get<Vec2>());
}

void to_json(nlohmann::json& j, const R4WarningEvent& w) {
  j = nlohmann::json{{"schemaVersion", w.schemaVersion},
                     {"type", w.type},
                     {"warningType", w.warningType},
                     {"riskState", w.riskState},
                     {"object", w.object},
                     {"geometry", w.geometry}};
}

void from_json(const nlohmann::json& j, R4WarningEvent& w) {
  j.at("schemaVersion").get_to(w.schemaVersion);
  j.at("type").get_to(w.type);
  j.at("warningType").get_to(w.warningType);
  j.at("riskState").get_to(w.riskState);
  j.at("object").get_to(w.object);
  j.at("geometry").get_to(w.geometry);
}

void to_json(nlohmann::json& j, const R4StateMessage& s) {
  j = nlohmann::json{{"schemaVersion", s.schemaVersion},
                     {"type", s.type},
                     {"seq", s.seq},
                     {"vehicles", s.vehicles}};
}

void from_json(const nlohmann::json& j, R4StateMessage& s) {
  j.at("schemaVersion").get_to(s.schemaVersion);
  j.at("type").get_to(s.type);
  j.at("seq").get_to(s.seq);
  j.at("vehicles").get_to(s.vehicles);
}

}  // namespace ada::contracts
